pinot-core/src/main/java/org/apache/pinot/core/accounting/PerQueryCPUMemAccountantFactory.java

```diff
@@ -814,14 +837,21 @@ void killAllQueries() {
 
         if (config.isOomKillQueryEnabled()) {
           int killedCount = 0;
+          Set<String> terminatedQueryIds = new HashSet<>();
           for (Map.Entry<Thread, CPUMemThreadLevelAccountingObjects.ThreadEntry> entry : _threadEntriesMap.entrySet()) {
             CPUMemThreadLevelAccountingObjects.ThreadEntry threadEntry = entry.getValue();
             CPUMemThreadLevelAccountingObjects.TaskEntry taskEntry = threadEntry.getCurrentThreadTaskStatus();
-            if (taskEntry != null && taskEntry.isAnchorThread()) {
-              threadEntry._errorStatus
-                  .set(new RuntimeException(String.format("Query killed due to %s out of memory!", _instanceType)));
-              taskEntry.getAnchorThread().interrupt();
-              killedCount += 1;
+            if (taskEntry != null && !terminatedQueryIds.contains(taskEntry.getQueryId())) {
+              if (_queryCancelCallbacks.containsKey(taskEntry.getQueryId())) {
+                _queryCancelCallbacks.get(taskEntry.getQueryId()).cancelQuery();
+                terminatedQueryIds.add(taskEntry.getQueryId());
+              } else if (taskEntry.isAnchorThread()) {
```

Comment:
Do we still need this older option?

Reply:
I'll keep it for now. 
* I removed it and reverted the change because the diff got too big. 
* Previously I had added checks to make sure anchor thread/parent context exists. Fixed NPE etc. 
I'll undo these changes esp. for MSE in a separate thread with unit tests. It will be safer that way.